File: maximum-subarray/maximum-subarray.dart
//Given a list of integers nums find the subarray
//that contains the highest value and return it.
//To do this iterate through nums each time adding to the current
//which is the value of the current subarray. If the subarray drops below 0
//then start a new subarray and continue on. Whenever you have a subarray with a value
//greater than that of the current max set max = current.
//The time complexity of this code is O(n) where n is the length
//of the list nums.
//Concept used: Sliding window
//Time: 340ms Beats: 100%
//Memory: 181.3MB Beats: 49.55%

class Solution {
  int maxSubArray(List<int> nums) {
      int max = -1000000;
      int current = 0;
      for(int i = 0 ; i < nums.length ; ++i){
          current += nums[i];
          if(current > max){
              max = current;
          }
          if(current < 0){
              current = 0;
          }
      }
      return max;
  }
}

File: maximum-subarray/maximum-subarray.cpp
#include <vector>
#include <iostream>
using namespace std;


//Leet Ratings
//       Speed  Memory
//Total  168ms  67.6MB    
//Beats  5.80%  93.55%


int maxSubArray(vector<int>& nums) {

    int size = nums.size();
    int i = 0;
    long int current_val = 0;
    long int highest = nums[0];
    while(i < size){
        current_val = nums[i];

        //This is necessary in case the final value of a list is the highest value.
        if(current_val > highest){
            highest = current_val;
        }
        int x = i + 1;
        while(x < size & current_val > 0){
            current_val += nums[x];
            if(current_val > highest){
                highest = current_val;
            }
            x += 1;
        }
        if(x == size){
            break;
        }

        i += 1;
        current_val = 0;
    }
    return highest;
} 

int main(){
        vector<int> vector = {10, -1, -10, 32, 3, -3};
        cout << maxSubArray(vector) << endl;
}
